File: csharp/Patterns/Observer.cs
using System;
using System.Collections.Generic;

namespace DesignPatterns.Patterns;

public static class Observer
{
    private interface ISubscriber
    {
        void Update(string news);
    }

    private class PhoneSubscriber : ISubscriber
    {
        private readonly string _name;

        public PhoneSubscriber(string name)
        {
            _name = name;
        }

        public void Update(string news)
        {
            Console.WriteLine($"{_name} got news: {news}");
        }
    }

    private class NewsPublisher
    {
        private readonly List<ISubscriber> _subscribers = new();

        public void Subscribe(ISubscriber subscriber) => _subscribers.Add(subscriber);
        public void Unsubscribe(ISubscriber subscriber) => _subscribers.Remove(subscriber);

        public void Publish(string news)
        {
            foreach (var subscriber in _subscribers)
            {
                subscriber.Update(news);
            }
        }
    }

    public static void Run()
    {
        Console.WriteLine("\nObserver");
        var publisher = new NewsPublisher();
        var alice = new PhoneSubscriber("Alice");
        var bob = new PhoneSubscriber("Bob");

        publisher.Subscribe(alice);
        publisher.Subscribe(bob);
        publisher.Publish("Ice cream truck is here!");
    }
}

File: python/patterns/observer.py
from abc import ABC, abstractmethod
from typing import List


class Subscriber(ABC):
    @abstractmethod
    def update(self, news: str) -> None:
        raise NotImplementedError


class PhoneSubscriber(Subscriber):
    def __init__(self, name: str) -> None:
        self.name = name

    def update(self, news: str) -> None:
        print(f"{self.name} got news: {news}")


class NewsPublisher:
    def __init__(self) -> None:
        self.subscribers: List[Subscriber] = []

    def subscribe(self, subscriber: Subscriber) -> None:
        self.subscribers.append(subscriber)

    def unsubscribe(self, subscriber: Subscriber) -> None:
        self.subscribers.remove(subscriber)

    def publish(self, news: str) -> None:
        for subscriber in self.subscribers:
            subscriber.update(news)


def run() -> None:
    print("\nObserver")
    publisher = NewsPublisher()
    alice = PhoneSubscriber("Alice")
    bob = PhoneSubscriber("Bob")

    publisher.subscribe(alice)
    publisher.subscribe(bob)
    publisher.publish("Ice cream truck is here!")
